src/asymmetric_encryption.cpp

```diff
@@ -0,0 +1,242 @@
+    : _data{data.begin(), data.end()}
+{
+}
+
+AsymmetricEncryption::CryptoData::CryptoData(const std::vector<uint8_t>& data)
+    : _data{data}
+{
+}
+
+AsymmetricEncryption::CryptoData& AsymmetricEncryption::CryptoData::operator=(
+                                                                    const std::vector<uint8_t>& data)
+{
+    _data=data;
+    return *this;
+}
+
+AsymmetricEncryption::CryptoData& AsymmetricEncryption::CryptoData::operator=(
+                                                                    const std::string& data)
+{
+    _data=std::vector<uint8_t>(data.begin(), data.end());
+    return *this;
+}
+
+std::string AsymmetricEncryption::CryptoData::toString() const {
+    return std::string(_data.begin(), _data.end());
+}
+
+std::vector<uint8_t> AsymmetricEncryption::CryptoData::toByteArray() const {
+    return _data;
+}
+
+std::string AsymmetricEncryption::CryptoData::toHex() const {
+    return utility::toHex(_data);
+}
+
+std::ostream& operator<<(std::ostream& os, const AsymmetricEncryption::CryptoData& data)
+{
+    os << data.toString();
+    return os;
+}
+
+std::vector<uint8_t> AsymmetricEncryption::encrypt(AsymmetricPublicKey key,
+                                                   const RSAPadding &pad,
+                                                   const CryptoData& message)
+{
+    SSL_RSA_OAEP_LABEL_Ptr label_copy{nullptr};
+    SSL_RSA_ENCRYPTION_DATA_Ptr encryptedMessage{nullptr};
+    size_t encryptedMessageLen{0};
+    int maxSize{0};
+
+    try {
+        const auto keyCtx = _EVP_PKEY_CTX_new(key.internal());
+
+        const auto paddingMode = pad.getPadding();
+
+        if (!keyCtx.get()) {
+            throw MoCOCrWException("Encryption context is empty");
+        }
+
+        switch(paddingMode) {
+            case RSAPaddingMode::PKCS1: {
+                maxSize = RSA_size(key.internal()->pkey.rsa) - c_pkcsMaxSizeSubtract;
+                break;
+            }
+            case RSAPaddingMode::OAEP: {
+                const auto oaepPaddingMode = static_cast<const OAEPPadding&>(pad);
+                maxSize = RSA_size(key.internal()->pkey.rsa) -
+                (2 * _EVP_MD_size(_getMDPtrFromDigestType(oaepPaddingMode.getHashingFunction())) - 2);
+                break;
+            }
+            case RSAPaddingMode::NONE: {
+                maxSize = RSA_size(key.internal()->pkey.rsa);
+                break;
+            }
+            default: throw MoCOCrWException("Unsupported encryption mode");
+        }
+
+        /* Validate message size */
+        if (static_cast<int>(message.toByteArray().size()) > maxSize) {
+            throw MoCOCrWException("Message too long for RSA key size");
+        }
+        _CRYPTO_malloc_init();
+        _EVP_PKEY_encrypt_init(keyCtx.get());
+        _EVP_PKEY_CTX_set_rsa_padding(keyCtx.get(), static_cast<int>(paddingMode));
+
+        if (RSAPaddingMode::OAEP == paddingMode) {
```

Comment:
Again, instead of requesting a type of padding, switching and casting, couldn't have something like `padding.configureContext(keyCtx)`